### PR TITLE
Manifest v1 0.0

### PR DESCRIPTION
**Pull request by jiqiang90**
TODO: 

- [x]  add validation for runner
- [x]   add validation for runner and ds in subql/validater

- [x] require at least one of genesis / chainId

- [x] add 1.0.0 to subq-migrate
- [x]  validation for runner semver version ranges
- [x] support 1.0.0 in subql-publish
- [ ] init (now yet, could support in next release with updated templates)
- [x]  inspect replace genesisHash in metadata impacts on hosted project 
- [ ] keep runner version in metadata, in order the subql/query can run validate

### CODE REVIEW

File: packages/cli/src/controller/init-controller.ts [hunk 1]
```diff
@@ -80,7 +80,7 @@ export async function readDefaults(projectPath: string): Promise<string[]> {
 
   const yamlPath = path.join(`${projectPath}`, `project.yaml`);
   const manifest = await fs.promises.readFile(yamlPath, 'utf8');
-  const currentProject = yaml.load(manifest) as ProjectManifestV0_0_1 | ProjectManifestV0_2_0;
+  const currentProject = yaml.load(manifest) as ProjectManifestV0_0_1 | SubstrateProjectManifestV0_2_0;
```
Comment (ianhe8x) [inline packages/cli/src/controller/init-controller.ts:83]: `subql init` shouldn't worry about the project's manifest impl. Can you only use the common interface?

File: packages/cli/src/controller/migrate-controller.ts [hunk 2]
```diff
@@ -87,7 +87,7 @@ export async function migrate(
   const manifestV0_2_0 = path.join(projectPath, MANIFEST_V_0_2_0);
 
   try {
-    const data = {} as ProjectManifestV0_2_0;
+    const data = {} as SubstrateProjectManifestV0_2_0;
```
Comment (ianhe8x) [inline packages/cli/src/controller/migrate-controller.ts:90]: same here, use the common interface.

File: packages/common-substrate/src/project/project.spec.ts [hunk 3]
```diff
@@ -28,6 +28,24 @@ describe('project.yaml', () => {
     expect(() => loadSubstrateProjectManifest(path.join(projectsDir, 'project_0.2.0_custom_ds.yaml'))).not.toThrow();
   });
 
+  it('can validate a v0.2.1 project.yaml with templates', () => {
```
Comment (ianhe8x) [inline packages/common-substrate/src/project/project.spec.ts:31]: What do you mean `with templates`. Can you add tests for terra manifest as a seperated one?

File: packages/common/src/project/versioned/v1_0_0/types.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+// Copyright 2020-2022 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import {Equals, IsString, Matches} from 'class-validator';
+import {RUNNER_REGEX} from '../../../constants';
+import {baseDataSource} from '../base';
+import {ProjectManifestV0_2_1} from '../v0_2_1';
+
+export interface runnerSpecs {
```
Comment (ianhe8x) [inline packages/common/src/project/versioned/v1_0_0/types.ts:9]: Capital case

File: packages/common-terra/src/project/versioned/v1_0_0/types.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2020-2022 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import {ProjectManifestV1_0_0} from '@subql/common';
+import {RuntimeDataSourceV0_3_0, CustomDatasourceV0_3_0} from '../v0_3_0';
+
+export interface TerraProjectManifestV1_0_0
```
Comment (ianhe8x) [inline packages/common-terra/src/project/versioned/v1_0_0/types.ts:7]: I don't think terra projects need a different Manifest interface.

File: packages/common/src/project/versioned/base.ts [hunk 6]
```diff
@@ -32,3 +32,21 @@ export abstract class ProjectManifestBaseImpl<D extends object> {
     }
   }
 }
+
+export interface baseDataSource {
```
Comment (ianhe8x) [inline packages/common/src/project/versioned/base.ts:36]: Capital case for all these three interfaces

Comment (ianhe8x) [inline packages/common/src/project/versioned/base.ts:36]:
```suggestion
export interface baseDataSource<T extends  baseMapping>{
```

File: packages/common/src/project/versioned/base.ts [hunk 7]
```diff
@@ -32,3 +32,21 @@ export abstract class ProjectManifestBaseImpl<D extends object> {
...
+  name?: string;
+  kind: string;
+  startBlock?: number;
+  mapping: baseMapping;
+}
+
+export interface baseMapping {
```
Comment (ianhe8x) [inline packages/common/src/project/versioned/base.ts:43]:
```suggestion
export interface baseMapping<T extends BaseHandler> {
```

File: packages/common/src/project/versioned/base.ts [hunk 8]
```diff
@@ -32,3 +32,21 @@ export abstract class ProjectManifestBaseImpl<D extends object> {
...
+  file: string;
+  handlers: baseHandler;
+}
+
+export interface baseHandler {
```
Comment (ianhe8x) [inline packages/common/src/project/versioned/base.ts:48]:
```suggestion
export interface baseHandler<T = Record<unknown, unknown>> { // T for filter
```

File: packages/common/src/project/versioned/v0_2_1/types.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2020-2022 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import {baseDataSource} from '@subql/common';
+import {ProjectManifestV0_2_0} from '../v0_2_0';
+
+export interface ProjectManifestV0_2_1<T, D extends object = baseDataSource> extends ProjectManifestV0_2_0<D> {
```
Comment (ianhe8x) [inline packages/common/src/project/versioned/v0_2_1/types.ts:7]: Need restriction for template type

File: packages/common-substrate/src/project/project.spec.ts (hunk 3 above)
Comment (jiqiang90) [reply packages/common-substrate/src/project/project.spec.ts:31]: As manifest 0.2.1 is for templates support, it should be able to validate a manifest with templates

File: packages/node/src/indexer/indexer.manager.ts [hunk 10]
```diff
@@ -362,21 +362,26 @@ export class IndexerManager {
...
-    if (!keyValue.genesisHash) {
+    if (!keyValue.chainId) {
       if (project) {
         await metadataRepo.upsert({
-          key: 'genesisHash',
+          key: 'chainId',
           value: project.networkGenesis,
         });
-      } else {
-        await metadataRepo.upsert({ key: 'genesisHash', value: genesisHash });
```
Comment (ianhe8x) [inline packages/node/src/indexer/indexer.manager.ts:373]: `else` block here was removed, does that mean when `!project`, no `chainId` will be inserted to _metadata?

**Review by ianhe8x — APPROVED** [review]

**Pull request merged**